README
#==============================================================
#
#  Name: README
#
#  Rem:	This is a driver for the Xycom 2440 32-channel optically
#		isolated digital input Industry Pack (IP) with interrupts.
#		This driver also supports the Acromag IP440
#		series digital input cards.
#
#		In order to deploy in a uniform fashion w/ the many other ip* modules,
#		this module has been renamed from xy2440 to ip440.
#		Support for the Xycom db, driver, and function names are all preserved
#		for compatibility, but ip440 variants have also been made available.
#
#  Auth: dd-mmm-yyyy, Andy Foster        (USERNAME)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        dd-mmm-yyyy, First Lastname     (USERNAME):
#          comment
#
#==============================================================
#
EPICS Verions:
--------------
This driver has currently been built and tested at SLAC with EPICS 3.14.8.2.
EPICS OSI support has been added to this driver by Till Strauman at SLAC.

Operating System:
-------------------
This driver has only been tested under RTEMS since osi support has been added.

How to use this module in your IOC
------------------
1.	Add the following lines in your iocTop/configure/RELEASE, using your desired IP440_MODULE_VERSION:
IP440_MODULE_VERSION=R2.3.0
IP440=$(EPICS_MODULES)/ip440/$(IP440_MODULE_VERSION)

2.	
For XyCom 2440, copy the following file to your iocTop/<iocName>/<iocName>App/Db directory:
	xy2440App/Db/xy2440.substitutions
For Acromag IP440, copy the following file to your iocTop/<iocName>/<iocName>App/Db directory:
	xy2440App/Db/ip440.substitutions
Customize as needed for your IOC

3.	
For XyCom 2440, add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += xy2440.db
For Acromag IP440, add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += ip440.db

4.	Add the following lines in your iocTop/<iocName>/<iocName>App/src/Makefile:
Note: Use either the ip440 variant or the Xy2440 variant here, but NOT both!
<iocName>_DBD += ip440.dbd
<iocName>_LIBS += ip440
   or
<iocName>_DBD += Xy2440.dbd
<iocName>_LIBS += Xy2440

Add the following lines in your iocTop/<iocName>/iocBoot/ioc<iocName>/st.cmd,
selecting the Xycom or Acromag as needed.
# Initialize IPAC carrier
# The 2nd parameter is the base I/O address set by the jumpers on your carrier board
# Make sure this address doesn't conflict with other boards.
# Note: ipac supports other carrier cards besides xy9660.
# See the ipac_carrier_t types in the ipac module for other options.
ipacAddCarrier(&xy9660,"0x0000")

# Initialize digital input module
# Set card to the carrier number, which is 0 for the first ipacAddCarrier() call and
# increments as other carrier cards are added.
# The slot numbers are 0..3 corresponding to slots A..D on the carrier card
# More than one digital input card can be used as long as the name and slot # differ.
# ip440Create( <name>, <card>, <slot>, <modeName>, <intHndlr>, <usrFunc>, <vector>, <event>, <debounce> )
ip440Create( "di0", 0, 0, "STANDARD", "LEVEL", 0x0, 0x80, 0x0, 0x00 )
# or for backward compatibility, you can use xy2440Create.  One call per card
xy2440Create( "di0", 0, 0, "STANDARD", "LEVEL", 0x0, 0x80, 0x0, 0x00 )

# Load the EPICS records
# For XyCom 2440:
dbLoadRecords("db/xy2440.db","CARD=di0")
# For Acromag IP440:
# dbLoadRecords("db/ip440.db","CARD=di0")

####################################################

EPICS PV Values:
----------------------
The default names assigned by the db template and substitution files
in this module are shown below.  These can be customized after you copy the
template and substitution files to your ioc.
------------------
PV Name:		Type:		Description:
di0:port0:in0	bi			Digital Input 1
di0:port0:in1	bi			Digital Input 2
di0:port0:in2	bi			Digital Input 3
di0:port0:in3	bi			Digital Input 4
di0:port0:in4	bi			Digital Input 5
di0:port0:in5	bi			Digital Input 6
di0:port0:in6	bi			Digital Input 7
di0:port0:in7	bi			Digital Input 8
di0:port1:in0	bi			Digital Input 9
di0:port1:in1	bi			Digital Input 10
di0:port1:in2	bi			Digital Input 11
di0:port1:in3	bi			Digital Input 12
di0:port1:in4	bi			Digital Input 13
di0:port1:in5	bi			Digital Input 14
di0:port1:in6	bi			Digital Input 15
di0:port1:in7	bi			Digital Input 16
di0:port2:in0	bi			Digital Input 17
di0:port2:in1	bi			Digital Input 18
di0:port2:in2	bi			Digital Input 19
di0:port2:in3	bi			Digital Input 20
di0:port2:in4	bi			Digital Input 21
di0:port2:in5	bi			Digital Input 22
di0:port2:in6	bi			Digital Input 23
di0:port2:in7	bi			Digital Input 24
di0:port3:in0	bi			Digital Input 25
di0:port3:in1	bi			Digital Input 26
di0:port3:in2	bi			Digital Input 27
di0:port3:in3	bi			Digital Input 28
di0:port3:in4	bi			Digital Input 29
di0:port3:in5	bi			Digital Input 30
di0:port3:in6	bi			Digital Input 31
di0:port3:in7	bi			Digital Input 32



RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem: This files describes release notes for
#       the Xycom 2440 driver and device support.
#
#==============================================================
#
R2.6.5   10-17-2022  V. Esposito
         Add build for rhel7, use IPAC R2.15-1.0.2


R2.6.3   08-12-2022  K. Luchini 
         Add build forlinuxRT-x86_64 in config/CONFIG_SITE 

R2.6.2:  Kukhee Kim - Oct 19, 2019
         Using IPAC R2.14-0.2.0

R2.6.1          Kukhee Kim, 10/16/2019
        use type-safe DSET and DRVET
        use ipac R2.14-0.2.0

R2.5.0-1.1.1:	  08-23-2017    Marcio Donadio
        Updated to ipac/R2.14-0.1.0

R2.5.0-1.1.0:     02-13-2017	Bruce Hill
	Updated to ipac/R2.8-1.2.1 for linux pciE support and removed obsolete dependency on miscUtils

R2.5.0-1.0.0:     02-10-2017	Bruce Hill
	Same as R2.4.0-1.2.0.  Updated version to match updated master version.

R2.5.0:           bhill, 2/10/2017
	Changed DTYP for ip440 to IP440.
	Was ACROMAG-IP440 in some files and ACROMAG-440 in others.
	New scheme matches other ip* modules

R2.4.0-1.2.0:     02-10-2017	Bruce Hill
	Changed DTYP from ACROMAG-IP440 to IP440 to match LCLS branch and other ip* modules

R2.4.0-1.1.0:     02-09-2017	Bruce Hill
	Restored xipIo code as it's no longer in most versions of the ipac module.
	Cleaned up some more build warnings re passing usrFunc as a char *

R2.4.0-1.0.0:     02-06-2017	Bruce Hill
	Merged in endian-fix.patch fix from R2.4.0-2.branch.
		Testing shows it works for both VME and linux, as the ip440 is an 8 bit module and
		Richard's 9378f4879 commit, "endianness support", changed all the I/O to 16 bit access.
		This allows the hardware on the carrier card and the host's PCI bridge chip to handle the byte order.
	Removed unneeded copies of xipIo code in ip445 which were cloned from the ipac module.
		This code is already available just by defining the IPAC macro in configure/RELEASE
	Applied ip440-signal-template-arch-req.patch to add archive and autosave template request files for ip440 signals.
	Added Acromag IP440 manual from pcds-trunk via "git cherry-pick 5bcd6bd"

R2.4.0:           bhill, 1/25/2017
	Added ip440Create, ip440.dbd, libip440.*, etc so this renamed ip440 module follows
	the same naming scheme as other ip modules.
	The prior xy2440Create, xy2440.dbd, libxy2440.* and such are still supported for
	backward compatibility.

R2.3.0:           bhill, 1/24/2017
				  Creating new release R2.3.0 to avoid confusion w/ prior pcds R2.2.* releases
                  Also updated to new git workflow base build support

xy2440-R1-0_1-0:  Build against ipac-R2-11_1-0 which builds for all targets

xy2440-R1-0-lcls6: Build with ipac-R2-11-lcls1
01/19/2012         Add xipIo.h to compile with ipac-R2-11-lcls1
                   Add xioIo.c file and copy xipIoParse() and strdupn() from drvXy9660.c (ipac-R2-8-lcls8).

xy2440-R1-0-lcls5: Build against EPICS R3-14-12 and use ipac-R2-8-lcls8

xy2440-R1-0-lcls4: Changes to decouple the version of EPICS BASE from the configure/RELEASE
                   Everything is essentially done in the RELEASE_SITE file now.

xy2440-R1-0-lcls3: Build with epics-R3-14-8-2-lcls6
                   Build with rtems-4.9.4

xy2440-R1-0-lcls2: Build with epics-R3-14-8-2-lcls5
                   Build with rtems-4.9.3

xy2440-R1-0-lcls1: Use naming convention for tag

xy2440-R2-0-0:   Build with epics-R3-14-8-2-lcls3
                 Build with rtems-4.9.1

xy2440-R1-0-7:   Change ptr to memory to volatile
                  struct config2440

xy2440-R1-0-6:   Build with base-R3-14-9-lcls1
                 Build with ipac-R2-8-lcls4
                 Replace CONFIG_APP with CONFIG_SITE
                 Add RELEASE_SITE

xy2440-R1-0-5:   Backout epics to base-R3-14-8-2-lcls2 (RTEMS 4.7.1)
                 Build with ipac-R2-8-lcls3

xy2440-R1-0-4:   Build with base-R3-14-9-lcls1 (RTEMS 4.7.1)
                 Build with extensions-R3-14-9
                 Build with ipac-R2-8-lcls2

xy2440-R1-0-3:   Build with base-R3-14-8-2-lcls2 (RTEMS 4.7.1)
                 Build with ipac-R2-8-lcls1

xy2440-R1-0-2:   Build with ipac-R1-0-2 instead of ipac-R1-0-1

xy2440-R1-0-1:   Change test databases

xy2440-R1-0-0:   Initial release

# End of file


xy2440App/src/xipIo.c
/*******************************************************************************

Project:
    IndustryPack Driver Interface for EPICS

File:
    xipIo.c

Copied xipIoParse() and strdupn() from original drvXy9660.c file
*******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include "xipIo.h"


char *strdupn( const char *ct, size_t n )
{
  char *duplicate;

  duplicate = (char *)malloc(n+1);
  if( !duplicate )
    return NULL;

  memcpy(duplicate, ct, n);
  duplicate[n] = '\0';

  return duplicate;
}

int xipIoParse( char *str, xipIo_t *ptr, char flag )
{
  char *name;
  char *end;

  if( str == NULL || ptr == NULL )
    return 1;

  while( !isalnum(*str) )
  {
    if( *str++ == '\0')
      return 1;
  }

  name = str;

  str = strpbrk(str, " ");
  if( str == NULL || *str == '\0' )
    return 1;

  ptr->name = strdupn(name, str - name);
  if( ptr->name == NULL )
    return 1;

  if( flag == 'A' )
  {
    end = strchr(str,'C');
    if( end )
    {
      str = end + 1;
      sscanf(str,"%d", &ptr->channel);
    }
    else
      return 1;
  }
  else if( flag == 'B' )
  {
    end = strchr(str,'P');
    if( end )
    {
      str = end + 1;
      sscanf(str,"%d", &ptr->port);
      end = strchr(str,'B');
      if( end )
      {
        str = end + 1;
        sscanf(str,"%d", &ptr->bit);
      }
      else
        return 1;
    }
    else
      return 1;
  }
  return(0);
}


xy2440App/src/devXy2440.c
/*******************************************************************************

Project:
    Gemini Multi-Conjugate Adaptive Optics Project

File:
    devXy2440.c

Description:
    EPICS Device Support for the XIP-2440-00x Industrial I/O Pack
    32-Channel Isolated Digital Input Module with Interrupts.

Author:
    Andy Foster <ajf@observatorysciences.co.uk>

Created:
      12th November 2002

Copyright (c) 2002 Andy Foster

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*******************************************************************************/

#define USE_TYPED_RSET

#include	<stdlib.h>
#include	<stdio.h>
#include	<string.h>

#include	"alarm.h"
#include	"cvtTable.h"
#include	"dbDefs.h"
#include        "devLib.h"
#include	"dbAccess.h"
#include        "recSup.h"
#include	"devSup.h"
#include	"dbScan.h"
#include	"link.h"
#include	"recGbl.h"
#include	"biRecord.h"
#include	"mbbiRecord.h"
#include	"mbbiDirectRecord.h"
#include	"drvXy2440.h"
#include	"xipIo.h"
#include        "epicsExport.h"

static long init_bi();
static long bi_ioinfo();
static long read_bi();

static long init_mbbi();
static long mbbi_ioinfo();
static long read_mbbi();

static long init_mbbiDirect();
static long mbbiDirect_ioinfo();
static long read_mbbiDirect();

typedef struct {
	long		number;
	DEVSUPFUN	report;
	DEVSUPFUN	init;
	DEVSUPFUN	init_record;
	DEVSUPFUN	get_ioint_info;
	DEVSUPFUN	read_write;
	} BINARYDSET;


BINARYDSET devBiXy2440         = {6, NULL, NULL, init_bi,   bi_ioinfo,   read_bi};
epicsExportAddress(dset, devBiXy2440);
BINARYDSET devMbbiXy2440       = {6, NULL, NULL, init_mbbi, mbbi_ioinfo, read_mbbi};
epicsExportAddress(dset, devMbbiXy2440);
BINARYDSET devMbbiDirectXy2440 = {6, NULL, NULL, init_mbbiDirect, mbbiDirect_ioinfo, 
                                  read_mbbiDirect};
epicsExportAddress(dset, devMbbiDirectXy2440);

/* Support Function */
static void handleError( void *prec, int *status, int error, char *errString, int pactValue );


static long init_bi(struct biRecord *pbi)
{
  xipIo_t        *pxip;
  int            status;
  unsigned char  handler;
  unsigned short value;
  void           *ptr;

  switch(pbi->inp.type) 
  {
    case(INST_IO):
      pxip = (xipIo_t *)malloc(sizeof(xipIo_t));
      if( !pxip )
      {
        handleError(pbi, &status, S_dev_noMemory, 
                    "devBiXy2440 (init_bi) malloc failed", TRUE);
      }
      else
      {
        /* Convert the address string into members of the xipIo structure */
        status = xipIoParse(pbi->inp.value.instio.string, pxip, 'B');
        if( status )
        {
          handleError(pbi, &status, S_xip_badAddress, 
                      "devBiXy2440 (init_bi) XIP address string format error", TRUE);
        }
        else
        {
          ptr = xy2440FindCard(pxip->name);
          if( ptr )
          {
            if( (pxip->port < 0) || (pxip->port >= MAXPORTS) )
            {
              handleError(pbi, &status, S_xy2440_portError,
                          "devBiXy2440 (init_bi) port out of range", TRUE);
            }
            else if( (pxip->bit  < 0) || (pxip->bit  >= MAXBITS)  )
            {
              handleError(pbi, &status, S_xy2440_bitError,
                          "devBiXy2440 (init_bi) bit out of range", TRUE);
            }
            else
            {
              /* Ask driver which interrupt handler do I have ? */
              xy2440WhichHandler( pxip->name, &handler );
              pxip->intHandler = handler;
              pbi->dpvt        = pxip;
              status           = xy2440Read(pxip->name, pxip->port, pxip->bit, BIT, &value, 0);
              if( status )
              {
                handleError(pbi, &status, S_xy2440_readError,
                            "devBiXy2440 (init_bi) error from xy2440Read", TRUE);
              }
              else
              {
                pbi->val  = value;
                pbi->rval = value;
              }
            }
          }
          else
          {
            handleError(pbi, &status, S_xy2440_cardNotFound,
                        "devBiXy2440 (init_bi) Card not found", TRUE);
          }
        }
      }
      break;

    default:
      handleError(pbi, &status, S_db_badField,
                  "devBiXy2440 (init_bi) illegal INP field", TRUE);
      break;
  }
  return(status);
}


static long bi_ioinfo( int cmd, struct biRecord *pbi, IOSCANPVT	*ppvt )
{
  xipIo_t *pxip;
  int     status;
  int     recType = BI;

  pxip   = (xipIo_t *)pbi->dpvt;
  status = xy2440GetIoScanpvt( pxip->name, pxip->port, pxip->bit, recType, 
                               pxip->intHandler, ppvt );
  if( status )
    handleError(pbi, &status, status, "devBiXy2440 (bi_ioinfo) error", FALSE);

  return(status);
}


static long read_bi(struct biRecord *pbi)
{
  xipIo_t        *pxip;
  int            status;
  unsigned short value;
	
  pxip   = (xipIo_t *)pbi->dpvt;
  status = xy2440Read(pxip->name, pxip->port, pxip->bit, BIT, &value, 0);
  if( status )
  {
    handleError(pbi, &status, S_xy2440_readError, "devBiXy2440 (read_bi) error", FALSE);
    recGblSetSevr(pbi,READ_ALARM,INVALID_ALARM);
    status = DO_NOT_CONVERT;
  }
  else
  {
    pbi->rval = value;
    status    = CONVERT;
  }
  return(status);
}


static long init_mbbi(struct mbbiRecord *pmbbi)
{
  xipIo_t        *pxip;
  int            status;
  unsigned char  handler;
  unsigned short value;
  void           *ptr;

  switch(pmbbi->inp.type) 
  {
    case(INST_IO):
      pxip = (xipIo_t *)malloc(sizeof(xipIo_t));
      if( !pxip )
      {
        handleError(pmbbi, &status, S_dev_noMemory,
                    "devMbbiXy2440 (init_mbbi) malloc failed", TRUE);
      }
      else
      {
        /* Convert the address string into members of the xipIo structure */
        status = xipIoParse(pmbbi->inp.value.instio.string, pxip, 'B');
        if( status )
        {
          handleError(pmbbi, &status, S_xip_badAddress,
                      "devMbbiXy2440 (init_mbbi) XIP address string format error", TRUE);
        }
        else
        {
          ptr = xy2440FindCard(pxip->name);
          if( ptr )
          {
            if( (pxip->port < 0) || (pxip->port >= MAXPORTS) )
            {
              handleError(pmbbi, &status, S_xy2440_portError,
                          "devMbbiXy2440 (init_mbbi) port out of range", TRUE);
            }
            else if( (pxip->bit  < 0) || (pxip->bit  >= MAXBITS)  )
            {
              handleError(pmbbi, &status, S_xy2440_bitError,
                          "devMbbiXy2440 (init_mbbi) bit out of range", TRUE);
            }
            else
            {
              /* Ask driver which interrupt handler do I have ? */
              xy2440WhichHandler( pxip->name, &handler );
              pxip->intHandler = handler;
              pmbbi->dpvt      = pxip;
              status           = xy2440Read(pxip->name, pxip->port, pxip->bit, NIBBLE, &value, 0);
              if( status )
              {
                handleError(pmbbi, &status, S_xy2440_readError,
                            "devMbbiXy2440 (init_mbbi) error from xy2440Read", TRUE);
              }
              else
              {
                pmbbi->val  = value & pmbbi->mask;
                pmbbi->rval = value & pmbbi->mask;
              }
            }
          }
          else
          {
            handleError(pmbbi, &status, S_xy2440_cardNotFound,
                        "devMbbiXy2440 (init_mbbi) Card not found", TRUE);
          }
        }
      }
      break;

    default:
      handleError(pmbbi, &status, S_db_badField,
                  "devMbbiXy2440 (init_mbbi) illegal INP field", TRUE);
      break;
  }
  return(status);
}


static long mbbi_ioinfo( int cmd, struct mbbiRecord *pmbbi, IOSCANPVT *ppvt )
{
  xipIo_t *pxip;
  int     status;
  int     recType = MBBI;

  pxip   = (xipIo_t *)pmbbi->dpvt;
  status = xy2440GetIoScanpvt( pxip->name, pxip->port, pxip->bit, recType, 
                               pxip->intHandler, ppvt );
  if( status )
    handleError(pmbbi, &status, status, "devMbbiXy2440 (mbbi_ioinfo) error", FALSE);

  return(status);
}


static long read_mbbi(struct mbbiRecord	*pmbbi)
{
  xipIo_t        *pxip;
  int            status;
  unsigned short value;
	
  pxip   = (xipIo_t *)pmbbi->dpvt;
  status = xy2440Read(pxip->name, pxip->port, pxip->bit, NIBBLE, &value, 0);
  if( status )
  {
    handleError(pmbbi, &status, S_xy2440_readError, "devMbbiXy2440 (read_mbbi) error", FALSE);
    recGblSetSevr(pmbbi,READ_ALARM,INVALID_ALARM);
    status = DO_NOT_CONVERT;
  }
  else
  {
    pmbbi->rval = value & pmbbi->mask;
    status      = CONVERT;
  }
  return(status);
}


static long init_mbbiDirect(struct mbbiDirectRecord *pmbbiDirect)
{
  xipIo_t        *pxip;
  int            status;
  unsigned char  handler;
  unsigned short value;
  void           *ptr;

  switch(pmbbiDirect->inp.type) 
  {
    case(INST_IO):
      pxip = (xipIo_t *)malloc(sizeof(xipIo_t));
      if( !pxip )
      {
        handleError(pmbbiDirect, &status, S_dev_noMemory,
                    "devMbbiDirectXy2440 (init_mbbiDirect) malloc failed", TRUE);
      }
      else
      {
        /* Convert the address string into members of the xipIo structure */
        status = xipIoParse(pmbbiDirect->inp.value.instio.string, pxip, 'B');
        if( status )
        {
          handleError(pmbbiDirect, &status, S_xip_badAddress,
                "devMbbiDirectXy2440 (init_mbbiDirect) XIP address string format error", TRUE);
        }
        else
        {
          ptr = xy2440FindCard(pxip->name);
          if( ptr )
          {
            if( (pxip->port < 0) || (pxip->port >= MAXPORTS) )
            {
              handleError(pmbbiDirect, &status, S_xy2440_portError,
                          "devMbbiDirectXy2440 (init_mbbiDirect) port out of range", TRUE);
            }
            else if( (pxip->bit  < 0) || (pxip->bit  >= MAXBITS)  )
            {
              handleError(pmbbiDirect, &status, S_xy2440_bitError,
                          "devMbbiDirectXy2440 (init_mbbiDirect) bit out of range", TRUE);
            }
            else
            {
              /* Ask driver which interrupt handler do I have ? */
              xy2440WhichHandler( pxip->name, &handler );
              pxip->intHandler  = handler;
              pmbbiDirect->dpvt = pxip;
              status            = xy2440Read(pxip->name, pxip->port, pxip->bit, WORD, &value, 0);
              if( status )
              {
                handleError(pmbbiDirect, &status, S_xy2440_readError,
                            "devMbbiDirectXy2440 (init_mbbiDirect) error from xy2440Read", TRUE);
              }
              else
              {
                pmbbiDirect->val  = value & pmbbiDirect->mask;
                pmbbiDirect->rval = value & pmbbiDirect->mask;
              }
            }
          }
          else
          {
            handleError(pmbbiDirect, &status, S_xy2440_cardNotFound,
                        "devMbbiDirectXy2440 (init_mbbiDirect) Card not found", TRUE);
          }
        }
      }
      break;

    default:
      handleError(pmbbiDirect, &status, S_db_badField,
                  "devMbbiDirectXy2440 (init_mbbiDirect) illegal INP field", TRUE);
      break;
  }
  return(status);
}


static long mbbiDirect_ioinfo( int cmd, struct mbbiDirectRecord *pmbbiDirect, 
                               IOSCANPVT *ppvt )
{
  xipIo_t *pxip;
  int     status;
  int     recType = MBBI_DIRECT;

  pxip   = (xipIo_t *)pmbbiDirect->dpvt;
  status = xy2440GetIoScanpvt( pxip->name, pxip->port, pxip->bit, recType, 
                               pxip->intHandler, ppvt );
  if( status )
    handleError(pmbbiDirect, &status, status, 
                "devMbbiDirectXy2440 (mbbiDirect_ioinfo) error", FALSE);

  return(status);
}


static long read_mbbiDirect( struct mbbiDirectRecord *pmbbiDirect )
{
  xipIo_t        *pxip;
  int            status;
  unsigned short value;
	
  pxip   = (xipIo_t *)pmbbiDirect->dpvt;
  status = xy2440Read(pxip->name, pxip->port, pxip->bit, WORD, &value, 0);
  if( status )
  {
    handleError(pmbbiDirect, &status, S_xy2440_readError, 
                "devMbbiDirectXy2440 (read_mbbiDirect) error", FALSE);
    recGblSetSevr(pmbbiDirect,READ_ALARM,INVALID_ALARM);
    status = DO_NOT_CONVERT;
  }
  else
  {
    pmbbiDirect->rval = value & pmbbiDirect->mask;
    status            = CONVERT;
  }
  return(status);
}


static void handleError( void *prec, int *status, int error, char *errString, int pactValue )
{
  struct dbCommon *pCommon;

  pCommon = (struct dbCommon *)prec;
  if( pactValue )
    pCommon->pact = TRUE;

  *status = error;
/*
  errlogPrintf("%s (%d): \"%s\"\n", pCommon->name, error, errString);
*/
  printf("%s (%d): \"%s\"\n", pCommon->name, error, errString);
}


xy2440App/src/drvXy2440.h
/*******************************************************************************

Project:
    Gemini Multi-Conjugate Adaptive Optics Project

File:
    drvXy2440.h

Description:
    Header file for the XIP-2440-00x Industrial I/O Pack
    32-Channel Isolated Digital Input Module with Interrupts.

Author:
    Andy Foster <ajf@observatorysciences.co.uk>

Created:
      12th November 2002

Modification:
      06-Nov-2008, K. Luchini (luchini@slac.stanford.edu)
        chagned ptr to map2440 in structure config2440 to volatile

Copyright (c) 2002 Andy Foster

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*******************************************************************************/

#ifndef INCdrvXy2440H
#define INCdrvXy2440H

/* Error numbers */

#ifndef M_xy2440
#define M_xy2440  (600 <<16)
#endif

#define S_xy2440_duplicateDevice    (M_xy2440| 1) /*Duplicate xy2440 device definition*/
#define S_xy2440_modeError          (M_xy2440| 2) /*Error in mode parameter*/
#define S_xy2440_intHandlerError    (M_xy2440| 3) /*Error in type of interrupt handler*/
#define S_xy2440_intConnectError    (M_xy2440| 4) /*Error from intConnect*/
#define S_xy2440_validateFailed     (M_xy2440| 5) /*Card not validated*/
#define S_xy2440_mallocFailed       (M_xy2440| 6) /*Malloc Failure*/
#define S_xy2440_portError          (M_xy2440| 7) /*Error in port number*/
#define S_xy2440_bitError           (M_xy2440| 8) /*Error in bit number*/
#define S_xy2440_readError          (M_xy2440| 9) /*Read error*/
#define S_xy2440_dataFlagError      (M_xy2440|10) /*Error in Data Flag*/
#define S_xy2440_cardNotFound       (M_xy2440|11) /*Card not found*/
#define S_xy2440_noInterrupts       (M_xy2440|12) /*Interrupts not supported in STANDARD mode*/
#define S_xy2440_invalidRecordType  (M_xy2440|13) /*Invalid record type for I/O scan support*/
#define S_xy2440_vectorInvalid      (M_xy2440|14) /*Interrupt vector number is invalid*/
#define S_xy2440_eventRegInvalid    (M_xy2440|15) /*Event register invalid*/
#define S_xy2440_debounceRegInvalid (M_xy2440|16) /*Debounce register invalid*/

/* EPICS Device Support return codes */

#define CONVERT        0
#define DO_NOT_CONVERT 2

#define BANK0   (unsigned char)0
#define BANK1   (unsigned char)1
#define BANK2   (unsigned char)2

#define MAXPORTS  4
#define MAXBITS   8

/* Data sizes that can be read */
#define BIT       0
#define NIBBLE    1
#define PORT      2
#define WORD      3

#define RESET     2   /* bit position in interrupt enable reg. */
#define INTEN     1   /* bit position in interrupt enable reg. */

#define STANDARD                 0        /* standard mode configuration */
#define ENHANCED                 1        /* enhanced mode configuration */

#define NOTUSED                  0        /* No interrupt handling       */
#define COS                      1        /* Change-of-state interrupt handling */
#define LEVEL                    2        /* Defined transition interrupt handling */

#define BI                       0        /* bi record type */
#define MBBI                     1        /* mbbi record type */
#define MBBI_DIRECT              2        /* mbbiDirect record type */

/* Parameter mask bit positions */

#define MASK                     2        /* write mask register */
#define EVCONTROL                4        /* event control register */
#define DEBCLOCK                 8        /* debounce clock register */
#define DEBCONTROL            0x10        /* debounce control register */
#define DEBDURATION           0x20        /* debounce duration register */
#define RESET_INTEN           0x40        /* interrupt enable register */
#define VECT                  0x80        /* interrupt vector register */

/* Memory Map for the Xy2440 Binary Input Module */

struct map2440
{
  struct
   {
     volatile unsigned short b_select;
   } port[8];
   volatile unsigned short nu0[7];                  
   volatile short ier;        /* interrupt enable register */
   volatile short nu1[7];     /* not used */
   volatile short ivr;        /* interrupt vector register */
};


/* Structure to hold the board's configuration information */

typedef void (*VOIDFUNPTR)();

struct config2440
{
    struct config2440 *pnext;                     /* to next device. Must be first member */
    char              *pName;                     /* Name to identify this card           */
    unsigned short    card;                       /* Number of IP carrier board           */
    unsigned short    slot;                       /* Slot number in carrier board         */
    volatile struct map2440    *brd_ptr;          /* base address of the input board      */
    unsigned short    param;                      /* parameter mask for configuring board */
    unsigned char     e_mode;                     /* enhanced operation flag              */
    unsigned char     mask_reg;                   /* output port mask register            */
    unsigned char     ev_control[2];              /* event control register               */
    unsigned char     deb_control;                /* debounce control register            */
    unsigned char     deb_duration;               /* debounce duration registers          */
    unsigned char     deb_clock;                  /* debounce clock select register       */
    unsigned char     enable;                     /* interrupt enable register            */
    unsigned char     vector;                     /* interrupt vector register            */
    volatile unsigned short    *id_prom;                   /* board ID Prom                        */
    unsigned char     ip_pos;                     /* IP under service position            */
    unsigned char     last_chan;                  /* last interrupt input channel number  */
    unsigned char     last_state;                 /* last state of the interrupt channel  */
    unsigned char     intHandler;                 /* interrupt handler flag               */
    VOIDFUNPTR        isr;                        /* Address of Interrupt Service Routine */
    VOIDFUNPTR        usrFunc;                    /* Address of user function             */
#ifndef NO_EPICS
    IOSCANPVT         biScan[MAXPORTS*MAXBITS];   /* One for each bit of each port, bi's  */
    IOSCANPVT         mbbiScan[MAXPORTS*MAXBITS]; /* All possible mbbi's                  */
    IOSCANPVT         mbbiDirectScan[MAXPORTS*MAXBITS]; /* All possible mbbiDirect's      */
#endif
};

#ifndef USE_TYPED_DRVET
int          xy2440Report( int interest );
int          xy2440Initialise( void );
int          ip440Report( int interest );
int          ip440Initialise( void );
#else
long         xy2440Report( int interest );
long         xy2440Initialise( void );
long         ip440Report( int interest );
long         ip440Initialise( void );
#endif

#ifndef NO_EPICS
int           xy2440GetIoScanpvt( char *name, unsigned char port, unsigned char point, 
                                  int mbbiscan, unsigned char intHandler, IOSCANPVT *ppvt );
#endif

int           xy2440Create( char *pName, unsigned short card, unsigned short slot,
                            char *modeName,
                            char *intHandlerName, VOIDFUNPTR usrFunc, short vector, 
                            short event, short debounce );
int           ip440Create(  char *pName, unsigned short card, unsigned short slot,
                            char *modeName,
                            char *intHandlerName, VOIDFUNPTR usrFunc, short vector, 
                            short event, short debounce );
void          xy2440SetConfig( char *pName, unsigned short card, unsigned short slot,
                               unsigned char mode,
                               unsigned char intHandler, VOIDFUNPTR usrFunc,
                               unsigned short vector, 
                               unsigned char event, unsigned char debounce, 
                               struct config2440 *pconfig );
void          xy2440Config( struct config2440 *pconfig );
unsigned char xy2440SelectBank( unsigned char newBank, struct config2440 *pconfig );
long          xy2440Read( char *name, short port, short bit, int readFlag,
                          unsigned short *pval, int debug );
void          xy2440COS( struct config2440 *pconfig );
void          xy2440LEVEL( struct config2440 *pconfig );
void          xy2440WhichHandler( char *name, unsigned char *handler );
void         *xy2440FindCard( char *name );

#endif  /* INCdrvXy2440H */
